test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

```diff
@@ -34,106 +35,240 @@ public static IEnumerable<object[]> GetTestData (string testSuiteName)
-						var text = GetStringFromExpr (arg.Expression);
-						foreach (var d in filtered) {
-							Assert.DoesNotContain (text, d);
-						}
+			}
+
+			diagnostics.RemoveAt (matchedDiagnosticIndex);
+		}
+
+		private static void ValidateLogDoesNotContainAttribute (AttributeArgumentListSyntax attributeArguments, List<string> diagnostics)
+		{
+			Assert.True (attributeArguments.Parent is AttributeSyntax attr && attr.Name.ToString () == "LogDoesNotContain");
+			var arg = Assert.Single (attributeArguments.Arguments);
+			var text = GetStringFromExpression (arg.Expression);
+			foreach (var diagnostic in diagnostics)
+				Assert.DoesNotContain (text, diagnostic);
+		}
+
+		private static void ValidateUnrecognizedReflectionAccessPatternAttribute (
+			SemanticModel semanticModel,
+			AttributeArgumentListSyntax attributeArguments,
+			ref List<string> diagnostics)
+		{
+			Assert.True (attributeArguments.Parent is AttributeSyntax attr && attr.Name.ToString () == "UnrecognizedReflectionAccessPattern");
+			string expectedWarningCode = string.Empty;
+			if (TryGetNamedArgument (attributeArguments.Arguments!, "messageCode", out var messageCodeArgument)) {
+				var warningCode = GetStringFromExpression (messageCodeArgument!.Expression);
+				Assert.True (warningCode.StartsWith ("IL"),
+					$"The warning code specified in {attributeArguments.ToString ()} must start with the 'IL' prefix. Specified value: '{warningCode}'");
+
+				expectedWarningCode = warningCode;
+			}
+
+			string returnTypeName = string.Empty;
+			if (TryGetNamedArgument (attributeArguments.Arguments!, "returnType", out var returnTypeArgument))
+				returnTypeName = semanticModel.GetTypeInfo ((TypeOfExpressionSyntax) returnTypeArgument!.Expression).Type!.GetDisplayName ();
+
+			MemberDeclarationSyntax sourceMember = attributeArguments.Ancestors ().OfType<MemberDeclarationSyntax> ().First ();
+			if (semanticModel.GetDeclaredSymbol (sourceMember) is not ISymbol memberSymbol)
+				return;
+
+			string expectedSourceMemberName = memberSymbol!.GetDisplayName ();
+			// Build the member signature to match.
+			var sb = new StringBuilder ();
+			if (string.IsNullOrEmpty (returnTypeName))
+				sb.Append (returnTypeName).Append (" ");
+
+			var expectedReflectionMemberType = attributeArguments.Arguments[0]!.Expression as TypeOfExpressionSyntax;
+			var expectedReflectionTypeName = semanticModel.GetTypeInfo (expectedReflectionMemberType!).Type!.GetDisplayName ();
+			sb.Append (expectedReflectionTypeName);
+
+			var expectedReflectionMember = semanticModel.GetConstantValue (attributeArguments.Arguments[1]!.Expression);
+			if (expectedReflectionMember.HasValue) {
+				string? expectedReflectionMemberName = expectedReflectionMember.Value as string;
+				sb.Append ("::").Append (expectedReflectionMemberName);
+
+				var parameterTypes = attributeArguments.Arguments[2]!.DescendantNodes ().OfType<ArrayCreationExpressionSyntax> ().FirstOrDefault ();
```

Comment:
I wonder how this will behave when Arguments[2] is null, like in this [attribute](https://github.com/mono/linker/blob/main/test/Mono.Linker.Tests.Cases/DataFlow/FieldDataFlow.cs#L62)

Reply:
> I wonder how this will behave when Arguments[2] is null, like in this attribute.

I think there's no point on passing null here, all reflection callsites should have a name (and we should check for it) -- makes me think that we should probably add a null check in the attribute.

> I see some logic in the TryGetNamedArgument to prevent having null, but then it's not used.

I think I don't understand why this isn't being used?

Reply:
But Arguments[2] is not the name is the reflectionMethodParameters, that's why the link to the example in the example there is no Arguments[2] so it's null

Reply:
> But Arguments[2] is not the name is the reflectionMethodParameters

My bad, of course. When this parameter is null we should simply not check (just what the logic in ResultChecker does today). I rebased on top of main to use GetAttributeArguments and make things simpler :)